rosenpass/src/app_server.rs

```diff
@@ -584,6 +604,55 @@ impl AppServer {
         }
     }
 
+    fn handle_msg_under_load(
+        &mut self,
+        endpoint: &Endpoint,
+        rx: &[u8],
+        tx: &mut [u8],
+    ) -> Result<crate::protocol::HandleMsgResult> {
+        let socket_addr = endpoint
+            .addresses()
+            .first()
+            .ok_or(anyhow::anyhow!("No socket address for endpoint"))?;
+
+        //TODO: Modify AppPeer to store endpoints to enable O(1) search (hash) for addresses
+        //If no endpoint matches address, return error
+        //Currently, there is no sensible mechanism to map "anonymous" endpoints to PeerPtr
+        let index = self
+            .peers
+            .iter()
+            .enumerate()
+            .find(|(_num, p)| {
+                if let Some(ep) = p.endpoint() {
+                    ep.addresses().contains(socket_addr)
+                } else {
+                    false
+                }
+            })
+            .ok_or(anyhow::anyhow!("Received message from unknown endpoint"))?
+            .0;
+
+        let mut len = 0;
+        let mut ip_addr_port = [0u8; 18];
+
+        match socket_addr.ip() {
```

Comment:
This encoding has some collisions and omits some data; i.e. there exist many pairs of ipv6 and ipv4 addresses with the same binary representation.

Lets

1. Include the mio socket token (i.e. the socket index)
2. Represent ipv4 addresses as [ipv6-mapped-ipv4](https://www.ibm.com/docs/en/zos/2.2.0?topic=addresses-ipv4-mapped-ipv6)
3. Include the [ipv6 scope id](https://doc.rust-lang.org/std/net/struct.SocketAddrV6.html#method.scope_id)

Can we factor this out into a special "struct HostIdentification { ... }"?

Reply:
Actually we already have a host identification: Endpoint::SocketBoundAddress.

Lets factor SocketBoundAddress into a struct `SocketBoundEndpoint` out.

try_recv can return the SocketBoundEndpoint and provide a nice `to_bytes` function as well as a `Display` implementation which can be used in the error handling code instead of manually printing the IP address there